File: CS.NodeJS.LoadTest.WebApp/webapp/logging/testResult.js
let eventResultModule = require('./eventResult.js');

module.exports = {

	_distinct: (acc, val) => {

		acc = Array.isArray(acc) ? acc : [];

		if (!acc.includes(val)) {
			acc.push(val);
		}

		return acc;
	},

	_clusterizeEvents: (eventResultArray) => {

		let eventSuperCluster = [];

		eventResultArray.forEach((e) => {

			let clusterDateTime = new Date(e.eventStart);
			clusterDateTime.setMilliseconds(0);

			let clusterIndex = eventSuperCluster.findIndex((cl) => {

				return cl.clusterDatetTime.getTime() == clusterDateTime.getTime();
			});

			if (clusterIndex > -1) {

				eventSuperCluster[clusterIndex].eventResultArray.push(e);
			}
			else {

				eventSuperCluster.push({
					clusterDatetTime: clusterDateTime,
					eventResultArray: [e]
				});
			}
		});

		return eventSuperCluster;
	},

	compileTestResult: (eventResultArray) => {

		let testResultArray = [];
		let testEvents = eventResultArray.filter((e) => { return e.test && e.test.testId; });

		if (testEvents.length < 1) {
			return testResultArray;
		}

		let testIdArray = testEvents
			.map((t) => { return t.test.testId; })
			.reduce(module.exports._distinct);

		let languageArray = testEvents
			.map((e) => { return e.test.language; })
			.reduce(module.exports._distinct);

		testIdArray.forEach((testId) => {
			languageArray.forEach((language) => {

				let testLanguageEvents = testEvents.filter((e) => {
					return e.test.testId == testId && e.test.language == language;
				});

				let testResult = {
					test: {
						testId: testId,
						language: language
					},

					metrics: eventResultModule.createEventResultClusterMetrics(testLanguageEvents),
					superCluster: []
				};

				let eventSuperCluster = module.exports._clusterizeEvents(testLanguageEvents);

				eventSuperCluster.forEach((cluster) => {

					let eventResultClusterMetrics = eventResultModule
						.createEventResultClusterMetrics(cluster.eventResultArray);

					testResult.superCluster.push(eventResultModule
						.createEventResultCluster(eventResultClusterMetrics, cluster.clusterDatetTime));
				});

				testResultArray.push(testResult);
			});
		});

		return testResultArray;
	}
}

File: DotNetCorePracticalView/CS.DotNetCore.LoadTest/src/CS.DotNetCore.LoadTest.WebApp/Logging/TestResult.cs
namespace CS.DotNetCore.LoadTest.WebApp.Logging
{
    using Newtonsoft.Json;
    using System;
    using System.Linq;
    using System.Collections.Generic;

    public class TestResult
    {
        [JsonProperty]
        internal EventResultClusterMetrics Metrics { get; private set; }

        [JsonProperty]
        internal Test Test { get; private set; }

        private List<EventResultCluster> _superCluster;

        [JsonProperty]
        internal IEnumerable<EventResultCluster> SuperCluster
        {
            get
            {
                return _superCluster;
            }

            private set
            {
                _superCluster = value.ToList();
            }
        }

        [JsonConstructor]
        private TestResult() { }

        internal TestResult(Test test, EventResultClusterMetrics metrics, IEnumerable<EventResultCluster> eventCluster)
        {
            Test = test;
            Metrics = metrics;
            SuperCluster = eventCluster == null ? new List<EventResultCluster>() : eventCluster;
        }

        private static Dictionary<DateTimeOffset, List<EventResult>> ClusterizeEvents(List<EventResult> eventResultColl)
        {
            var eventSuperCluster = new Dictionary<DateTimeOffset, List<EventResult>>();

            foreach (var eventResult in eventResultColl)
            {
                var clusterKey = new DateTimeOffset
                (
                    eventResult.EventStart.Year,
                    eventResult.EventStart.Month,
                    eventResult.EventStart.Day,
                    eventResult.EventStart.Hour,
                    eventResult.EventStart.Minute,
                    eventResult.EventStart.Second,
                    eventResult.EventStart.Offset
                );

                if (!eventSuperCluster.ContainsKey(clusterKey))
                {
                    eventSuperCluster.Add(clusterKey, new List<EventResult>());
                }

                eventSuperCluster[clusterKey].Add(eventResult);
            }

            return eventSuperCluster;
        }

        internal static List<TestResult> CompileTestResults(IEnumerable<EventResult> eventResultColl)
        {
            //resolving testid and languages from event results
            var testEvents = eventResultColl.Where(e => e.Test != null && e.Test.TestId != null);
            var testIdColl = testEvents.Select(e => e.Test.TestId).Distinct();

            var testLanguageColl = testEvents.Select(e => e.Test.Language).Distinct();
            var testResultColl = new List<TestResult>();

            foreach (var testId in testIdColl)
            {
                foreach (var language in testLanguageColl)
                {
                    //resolving events from a specific id and language
                    var testLanguageEvents = testEvents
                        .Where(e => e.Test.TestId == testId && e.Test.Language == language)
                        .ToList();

                    //calculating metrics from all events in the test
                    var testResult = new TestResult()
                    {
                        Test = new Test(testId, language),
                        Metrics = new EventResultClusterMetrics(testLanguageEvents),
                    };

                    //clusterizing events in the test
                    var eventSuperCluster = ClusterizeEvents(testLanguageEvents);

                    testResult._superCluster = eventSuperCluster.Select
                    (
                        sc => new EventResultCluster(new EventResultClusterMetrics(sc.Value), sc.Key)
                    )
                    .ToList();

                    testResult._superCluster.Sort((c1, c2) => { return c1.ClusterDateTime.CompareTo(c2.ClusterDateTime); });
                    testResultColl.Add(testResult);
                }
            }

            return testResultColl;
        }
    }
}
